Proposals/20200511-HVPA-design.md

```diff
@@ -191,6 +191,43 @@ New Requests: 225m / 2.7Gi with 2 replicas, because 1.8Gi lands us in the “sec
 Notice: Old and new CPU would even land us in bucket 1, but memory brings us only down to bucket 2, so we go to bucket 2 with 2 replicas.
 Reasoning: We don’t know what the cache percentage is and if it is very low, the remaining replicas will need even more memory to cope with the requests. If not, VPA will eventually scale down.
 
+#### Providing stability at bucket boundaries (TBD):
+Given that the **buckets are decided based on total resource allocation, and NOT the actual replica sizes**, if the total memory recommendation in the above example fluctuates around 2GB, then HVPA controller would be switching repeatedly between 1st and 2nd bucket.
+To prevent this some of the options are discussed below:
+
+**Option 1**
+
+Introduce new fields which will define lower bounds of each bucket - `minCpu` and `minMemory`.
+The values for these fields can be set such that there is an overlap between 1st bucket and 2nd bucket in terms of **aggregate resource value**. For eg., considering only 1st 2 buckets for brevity:
+```yaml
+spec:
+  scaleIntervals:
+  - minCPu: 125m,
+    minMemory: 500M,
+    maxCpu: "500m",      # Total max: 500m * 1 = 500m
+    maxMemory: "2G",     # Total max: 2G * 1 = 2G
+    maxReplicas: "1",
+  - minCPu: 200m,        # Total min: 200m * 2 = 400m (less than max of previous bucket)
+    minMemory: 900M,     # Total min: 900M * 2 = 1800M (less than max of previous bucket)
+    maxCpu: "1",
+    maxMemory: "4G",
+    maxReplicas: "2",
+```
+If we are already in 2nd bucket with 2 replicas such that *total* resource allocation is: 2.2G memory and 600m core CPU, and new *total* recommendations are 1.9G memory and 450m CPU, which would put us in both the buckets, then scaling would happen according to 2nd bucket, i.e., replica count would be kept unchanged to 2, and cpu and memory will be scaled down accordingly.
+
+**Option 2**
+
+Depend upon existing `minChange` fields as explained below.
+
+`minChange` fields are used to block any scaling if the difference between new recommendations and current resource allocations is less than the `minChange` provided in the spec. This is used to prevent scaling if recommendations fluctuate around any value. Even if the recommendations fluctuate around a bucket boundary, HVPA would not switch buckets if the difference is less than values provided under `minChange` for CPU and memory.
+For eg.,
+```yaml
+minChange:
+  - memory: 400M
+    cpu: 200m
+```
+If we are already in 2nd bucket, i.e., *total* resource allocation is: 2.2G memory and 600m core CPU, and new *total* recommendations are 1.9G memory and 450m CPU, which would put us in both the buckets, then scaling would not happen at all, because the differences between recommendations and current allocations are 300M and 150m, which are less than `minChange` values of 400M and 200m respectively.
+
```

Comment:
One more option could be to introduce a new field like `bucketOverlap` either per bucket or one field for all buckets. Just comparing with `minChange` option, with the `minChange` option we would be using it for two technically different purposes. Having a separate field might be more flexible.

This is just a rudimentary crack at a kind of _smart_ formula that @vlerenc mentioned in offline discussion. There might be other more sophisticated options.

Reply:
added these options in https://github.com/gardener/hvpa-controller/pull/71/commits/33aef9bd412492ab2b43623ef7cdd8fce0f65e8d.
I have added pros and cons too. Please let me know if there are some more that I might have missed.
thanks.